genesyscloud/architect_grammar/resource_genesyscloud_architect_grammar.go

```diff
@@ -0,0 +1,189 @@
+				FileType: language.VoiceFileMetadata.FileType,
+			}
+			err = proxy.uploadGrammarLanguageFile(*grammar.Id, *language.Language, language.VoiceFileMetadata.FileName, &uploadRequest, "voice")
+			if err != nil {
+				return diag.Errorf("Failed to upload language voice file: %s", err)
+			}
+		}
+
+		// Upload grammar dtmf file
+		if language.DtmfFileMetadata != nil && language.DtmfFileMetadata.FileName != nil {
+			uploadRequest := platformclientv2.Grammarfileuploadrequest{
+				FileType: language.DtmfFileMetadata.FileType,
+			}
+			err = proxy.uploadGrammarLanguageFile(*grammar.Id, *language.Language, language.DtmfFileMetadata.FileName, &uploadRequest, "dtmf")
+			if err != nil {
+				return diag.Errorf("Failed to upload language dtmf file: %s", err)
+			}
+		}
+	}
+
+	d.SetId(*grammar.Id)
+	log.Printf("Created Architect Grammar %s", *grammar.Id)
+	return readArchitectGrammar(ctx, d, meta)
+}
+
+// readArchitectGrammar is used by the architect_grammar resource to read an architect grammar from genesys cloud.
+func readArchitectGrammar(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*gcloud.ProviderMeta).ClientConfig
+	proxy := newArchitectGrammarProxy(sdkConfig)
+
+	log.Printf("Reading Architect Grammar %s", d.Id())
+
+	return gcloud.WithRetriesForRead(ctx, d, func() *retry.RetryError {
+		grammar, respCode, getErr := proxy.getArchitectGrammarById(ctx, d.Id())
+
+		if getErr != nil {
+			if gcloud.IsStatus404ByInt(respCode) {
+				return retry.RetryableError(fmt.Errorf("Failed to read Architect Grammar %s: %s", d.Id(), getErr))
+			}
+			return retry.NonRetryableError(fmt.Errorf("Failed to read Architect Grammar %s: %s", d.Id(), getErr))
+		}
+
+		//cc := consistency_checker.NewConsistencyCheck(ctx, d, meta, ResourceArchitectGrammar())
+
+		resourcedata.SetNillableValue(d, "name", grammar.Name)
+		resourcedata.SetNillableValue(d, "description", grammar.Description)
+		resourcedata.SetNillableValueWithInterfaceArrayWithFuncWithState(d, "languages", grammar.Languages, flattenGrammarLanguages)
+		
+		log.Printf("Read Architect Grammar %s %s", d.Id(), *grammar.Name)
+		//return cc.CheckState()
+		return nil
+	})
+}
+
+// updateArchitectGrammar is used by the architect_grammar resource to update an architect grammar in Genesys Cloud
+func updateArchitectGrammar(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*gcloud.ProviderMeta).ClientConfig
+	proxy := newArchitectGrammarProxy(sdkConfig)
+
+	architectGrammar := getArchitectGrammarFromResourceData(d)
+
+	// Update grammar
+	log.Printf("Updating Architect Grammar %s", *architectGrammar.Name)
+	grammar, err := proxy.updateArchitectGrammar(ctx, d.Id(), &architectGrammar)
+	if err != nil {
+		return diag.Errorf("Failed to update grammar: %s", err)
+	}
+
+	// Update each language associated with the grammar
+	for _, language := range *architectGrammar.Languages {
```

Comment:
Might be better to move all this logic inside the `for` statement over to the proxy file. So you would just call the update function here, and the proxy would take care of everything else. Same for the create